language/stdlib/modules/libra_system.move

```diff
@@ -186,69 +196,186 @@ module LibraSystem {
+        _  = Vector::swap_remove(
+            &mut discovery_set_ref.discovery_set,
+            to_remove_index
+        );
+
+        reconfigure_();
+        emit_discovery_set_change();
+    }
+
+    public fun rotate_consensus_pubkey(consensus_pubkey: vector<u8>) acquires ValidatorSet {
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        let account_address = Transaction::sender();
+
+        // Ensure that this address is already a validator
         Transaction::assert(
-            !Vector::contains(
-                &validator_set_ref.removals,
-                &account_address,
-            ),
-            22
+            is_validator_(&account_address, &validator_set_ref.validators),
+            21
         );
 
-        // Add to removals
-        Vector::push_back(
-            &mut validator_set_ref.removals,
-            account_address,
+        ValidatorConfig::rotate_consensus_pubkey(consensus_pubkey);
+
+        let validator_index = get_validator_index(
+            &validator_set_ref.validators,
+            account_address
+        );
+
+        if(copy_validator_info(Vector::borrow_mut(&mut validator_set_ref.validators, validator_index))) {
+            reconfigure_();
+        }
+    }
+
+    public fun rotate_validator_network_identity_pubkey(
+        validator_network_identity_pubkey: vector<u8>
+    ) acquires DiscoverySet {
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        let account_address = Transaction::sender();
+
+        ValidatorConfig::rotate_validator_network_identity_pubkey(validator_network_identity_pubkey);
+
+        let validator_index = get_discovery_index(
+            &discovery_set_ref.discovery_set,
+            account_address
+        );
+
+        if(copy_discovery_info(Vector::borrow_mut(&mut discovery_set_ref.discovery_set, validator_index))) {
+            emit_discovery_set_change();
+        }
+    }
+
+    public fun rotate_validator_network_address(
+        validator_network_address: vector<u8>
+    ) acquires DiscoverySet {
+        let discovery_set_ref = borrow_global_mut<DiscoverySet>(0xD15C0);
+        let account_address = Transaction::sender();
+
+        ValidatorConfig::rotate_validator_network_address(validator_network_address);
+
+        let validator_index = get_discovery_index(
+            &discovery_set_ref.discovery_set,
+            account_address
+        );
+
+        if(copy_discovery_info(Vector::borrow_mut(&mut discovery_set_ref.discovery_set, validator_index))) {
+            emit_discovery_set_change();
+        }
+    }
+
+    fun reconfigure_() acquires ValidatorSet {
+        let validator_set_ref = borrow_global_mut<ValidatorSet>(0x1D8);
+        let current_block_time = LibraTimestamp::now_microseconds();
+
+        validator_set_ref.last_reconfiguration_time = current_block_time;
+
+        // Only one reconfiguration allowed per block for now. We might want to increase the threshold here.
```

Comment:
What is the motivating for increasing the threshold? I think I would say something like:

"Ensure that there is at most one reconfiguration per transaction. This ensures that there is a 1-1 correspondence between system reconfigurations and emitted `ReconfigurationEvents`."

Reply:
Also: can you add a test that trips this assert?

Reply:
This is actually tested in one of your comment below.

Reply:
Looks like this tests tripping this via two `add`s, which is good. But for completeness, it would also be useful to test other permutations like:
- Double rotation
- Add + remove
- Discovery set change + validator set change

Reply:
Actually DiscoverySet change is not a reconfiguration according to the existing module. I would argue that we should not change this behavior just for this PR and I'll work with @phlip9 on the actual implementation details of discovery set.